sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GlobalAddressResolver.java

```diff
@@ -95,61 +96,67 @@ public GlobalAddressResolver(
+
+                            return this.routingMapProvider.tryGetOverlappingRangesAsync(
+                                            null,
+                                            collection.getResourceId(),
+                                            PartitionKeyInternalHelper.FullRange,
+                                            true,
+                                            null)
+                                    .map(valueHolder -> {
+
+                                        if (valueHolder == null || valueHolder.v == null || valueHolder.v.size() == 0) {
+                                            logger.warn(
+                                                    "There is no pkRanges found for collection {}, no connections will be opened",
+                                                    collection.getResourceId());
+                                            return new ArrayList<PartitionKeyRangeIdentity>();
+                                        }
+
+                                        return valueHolder.v
+                                                .stream()
+                                                .map(pkRange -> new PartitionKeyRangeIdentity(collection.getResourceId(), pkRange.getId()))
+                                                .collect(Collectors.toList());
+                                    })
+                                    .flatMapMany(pkRangeIdentities -> this.openConnectionsAndInitCachesInternal(collection, pkRangeIdentities, proactiveContainerInitConfig));
+                        }));
     }
 
     private Flux<OpenConnectionResponse> openConnectionsAndInitCachesInternal(
             DocumentCollection collection,
-            List<PartitionKeyRangeIdentity> partitionKeyRangeIdentities) {
-
-        // Currently, we will only open connections to current read region
-        return Flux.just(this.endpointManager.getReadEndpoints().stream().findFirst())
-                .flatMap(readEndpointOptional -> {
-                    if (readEndpointOptional.isPresent()) {
-                        if (this.addressCacheByEndpoint.containsKey(readEndpointOptional.get())) {
-                            return this.addressCacheByEndpoint.get(readEndpointOptional.get())
-                                        .addressCache
-                                        .openConnectionsAndInitCaches(collection, partitionKeyRangeIdentities);
+            List<PartitionKeyRangeIdentity> partitionKeyRangeIdentities,
+            CosmosContainerProactiveInitConfig proactiveContainerInitConfig
+    ) {
+
+        if (proactiveContainerInitConfig.getNumProactiveConnectionRegions() > 0) {
+            return Flux.fromStream(this.endpointManager.getReadEndpoints().stream())
+                    .take(proactiveContainerInitConfig.getNumProactiveConnectionRegions())
+                    .flatMap(readEndpoint -> {
+                        if (this.addressCacheByEndpoint.containsKey(readEndpoint)) {
```

Comment:
I kind feel we should do the region opening sequentially.  For example, i want to open 2 regions, and if there is a resource limitation that i can not open all connections, then I would prefer to open more in the 1st preferred region than the 2nd region. 

 Definitely no-blocking for them (and probably I have worried too much). But in the future, when we define the failure threshold, it may matters.

Reply:
it is a trade-off - when partition count is relatively low (<= 1,000 partitions) this seems to have better perf - somewhere between 1,000 and 5,000 partitions is the turning point where proactively opening connections to multiple regions doesn't make much sense (5,000 * 4 * 2 - 40,000 ephemeral ports) - I think I would just explain that in the docs that proactively opening connections to remote regions is a trade-off and what to watch out for.